Add support for use pyproject.toml (#1227)
django-wiki/django-wiki: docs/development/hatch.rst

pyproject.toml and hatch
========================

> Time ago django-wiki core dev's introduced the possibility to start using the
> new ``pyproject.toml`` and drop the old ``setup.py`` proposed in the issue
> :url-issue:`1199`, this was due to The Python packaging ecosystem has
> standardized on the interface for build backends
> (`PEP 517 <https://peps.python.org/pep-0517/>`_/`PEP 660 <https://peps.python.org/pep-0660/>`_)
> and the format for metadata declaration (`PEP 621 <https://peps.python.org/pep-0621/>`_/`PEP 631 <https://peps.python.org/pep-0631/>`_).
> As a result, the execution of ``setup.py`` files is now `deprecated <https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html>`_.

Reviewer: ```suggestion
Initially, django-wiki core devs opted to use the
new ``pyproject.toml`` and drop the old ``setup.py`` (see
:url-issue:`1199`). The Python packaging ecosystem has
standardized on the interface for build backends
(`PEP 517 <https://peps.python.org/pep-0517/>`_/`PEP 660 <https://peps.python.org/pep-0660/>`_)
and the format for metadata declaration (`PEP 621 <https://peps.python.org/pep-0621/>`_/`PEP 631 <https://peps.python.org/pep-0631/>`_).
As a result, the execution of ``setup.py`` files is now `deprecated <https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html>`_.
```

  |         |         | py3.8-3.1  | flake8<5.1,>=3.7          | no-cov      |
  |         |         | py3.8-3.2  | pre-commit                | test        |
  |         |         | py3.9-3.0  | pytest-cov                |             |
  |         |         | py3.9-3.1  | pytest-django             |             |
  |         |         | py3.9-3.2  | pytest-pythonpath         |             |
  |         |         | py3.10-3.2 | pytest<7.3,>=6.2.5        |             |
  |         |         | py3.7-4.0  |                           |             |
  |         |         | py3.8-4.0  |                           |             |
  |         |         | py3.9-4.0  |                           |             |
  +---------+---------+------------+---------------------------+-------------+

We have 4 different environments declared in the configuration file, each one
has his own purpose::

* ``default``: The development environment for django-wiki.

* ``test``: where we ensure that the code works on different Django and Python versions.

* ``docs``: Used for generate the page you're reading at this moment.

* ``transifex``: Used only for the translation side of the project.

> We use heavily on the scripts functionality provided by ``hatch`` (`read more <https://hatch.pypa.io/latest/environment/#scripts>`_)

Reviewer: ```suggestion
We center around the entrypoints provided by ``hatch`` (`read more <https://hatch.pypa.io/latest/environment/#scripts>`_)
```

> that's why we have included some commands that will make the development
> process more easier. Some commands are only available at certain environments,

Reviewer: ```suggestion
that's why we have documented commands that make development easier.
Some commands are only available in certain environments,
```

so for example at the ``transifex`` environment you see ``pull`` and ``push``
commands that are not present in any other environment declared above. For
executing the command you have to follow this simple formula::

  $ hatch run <environment name>:<command name>

Then applied to the ``push`` command on the ``transifex`` environment will be::

  $ hatch run transifex:pull

> You can use the same logic for execute the available commands in the app, but

Reviewer: ```suggestion
You can use the same logic to execute the available commands in the app, but
```

  * ``docs:html``

  * ``docs:dirhtml``

  * ``docs:singlehtml``

  * ``docs:pickle``

  * ``docs:json``

  * ``docs:htmlhelp``

  * ``docs:qthelp``

  * ``docs:devhelp``

  * ``docs:epub``

  * ``docs:latex``

  * ``docs:latexpdf``

  * ``docs:text``

  * ``docs:man``

  * ``docs:texinfo``

  * ``docs:info``

  * ``docs:gettext``

  * ``docs:changes``

  * ``docs:link-check2``

  * ``docs:doctest``

* ``docs:build``: Generate the documentation in HTML format.

* ``docs:link-check``: Checks for external links across the documentation.

We hope that this document helps you to understand more about the development
process, if something is not clear please open an issue.

FAQ
---

1. **Whats the difference between test and test:all?**

   When you execute ``hatch run test`` this will check your changes in the
   active environment, this means it will run over an specific Python version
   and an specific Django Version; in the other hand ``test:all`` will run the
   test suite in the whole matrix of the supported versions of Python and Django.

2. **What Python Versions are currently used?**

   Right now, ``django-wiki`` runs on Python 3.7.X until 3.10.X and for Django

>    it runs in version 2.2 until version 4.0.

Reviewer: I think this section should be removed since it's mentioned in the main README.. so the above information will be hard to maintain. Likely, it will be inaccurate 1 year from now :)
